File: apps/ios/Padel/Design/Components.swift
@preconcurrency import PadelShared
import SwiftUI

struct PrimaryButtonStyle: ButtonStyle {
    @Environment(\.isEnabled) private var isEnabled

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(.geist(Tokens.FontSize.base, weight: .semibold))
            .frame(maxWidth: .infinity, minHeight: Tokens.touchTarget + 4)
            .foregroundStyle(Palette.primaryForeground)
            .background(Palette.primary.opacity(isEnabled ? 1 : 0.35), in: RoundedRectangle(cornerRadius: Tokens.Radius.md))
            .opacity(configuration.isPressed ? 0.85 : 1)
            .animation(.easeOut(duration: Tokens.motionFast), value: configuration.isPressed)
            .sensoryFeedback(.impact(weight: .light), trigger: configuration.isPressed) { _, pressed in pressed }
    }
}

struct SecondaryButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(.geist(Tokens.FontSize.base, weight: .semibold))
            .frame(maxWidth: .infinity, minHeight: Tokens.touchTarget + 4)
            .foregroundStyle(Palette.foreground)
            .background(configuration.isPressed ? Palette.muted : Palette.background, in: RoundedRectangle(cornerRadius: Tokens.Radius.md))
            .overlay(RoundedRectangle(cornerRadius: Tokens.Radius.md).stroke(Palette.border))
            .animation(.easeOut(duration: Tokens.motionFast), value: configuration.isPressed)
    }
}

/// Bordered surface card, the web's `Card`.
struct CardModifier: ViewModifier {
    var padding: CGFloat = Tokens.Space.s4

    func body(content: Content) -> some View {
        content
            .padding(padding)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Palette.surface, in: RoundedRectangle(cornerRadius: Tokens.Radius.lg))
            .overlay(RoundedRectangle(cornerRadius: Tokens.Radius.lg).stroke(Palette.border))
    }
}

extension View {
    /// iPad: a readable centred column, like the web's max-w container.
    func readableWidth() -> some View { frame(maxWidth: 720).frame(maxWidth: .infinity) }

    func card(padding: CGFloat = Tokens.Space.s4) -> some View { modifier(CardModifier(padding: padding)) }
}

struct SectionTitle: View {
    let text: String
    var trailing: String?

    init(_ text: String, trailing: String? = nil) {
        self.text = text
        self.trailing = trailing
    }

    var body: some View {
        HStack {
            Text(text).font(.geist(Tokens.FontSize.lg, weight: .semibold, relativeTo: .headline))
            Spacer()
            if let trailing {
                Text(trailing).font(.geist(Tokens.FontSize.sm)).foregroundStyle(Palette.mutedForeground)
            }
        }
        .foregroundStyle(Palette.foreground)
    }
}

/// *Live · view only*, *Synced*, *Offline — 3 changes waiting* (FR-2.10).
struct StatusLine: View {
    let game: LocalGame

    var body: some View {
        HStack(spacing: Tokens.Space.s2) {
            Circle().fill(dotColor).frame(width: 7, height: 7)
            Text(text)
        }
        .font(.geist(Tokens.FontSize.sm, relativeTo: .subheadline))
        .foregroundStyle(Palette.mutedForeground)
        .accessibilityElement(children: .combine)
    }

    private var text: String {
        var parts = [game.game.status == .done ? "Finished" : "Live"]
        if !game.canEdit { parts.append("view only") }
        switch game.syncState {
        case .localOnly: parts.append("on this phone — shares when online")
        case .pending: parts.append("offline · \(game.pending.count) change\(game.pending.count == 1 ? "" : "s") waiting")
        default: break
        }
        return parts.joined(separator: " · ")
    }

    private var dotColor: Color {
        if game.game.status == .done { return Palette.mutedForeground }
        return game.syncState == .synced ? Palette.foreground : Palette.mutedForeground
    }
}

extension ModeInfo {
    /// Same mapping as the web mode cards.
    var icon: PadelIcon {
        switch id {
        case .americano: .rotate
        case .teamAmericano, .teamMexicano: .pair
        case .mexicano: .ladder
        case .mixicano: .mixed
        case .beatTheBox: .box
        case .upAndDown, .teamUpAndDown: .updown
        default: .racket
        }
    }
}

extension PadelIcon {
    func view(_ size: CGFloat = Tokens.iconSize) -> some View {
        image.resizable().scaledToFit().frame(width: size, height: size)
    }
}

extension GameState {
    func name(_ id: String) -> String { players.first { $0.id == id }?.name ?? "?" }
    func team(_ ids: [String]) -> String { ids.map(name).joined(separator: " & ") }
    var mode: ModeInfo { ModesKt.modeInfo(id: settings.mode) }
    var roundLabel: String {
        let n = Int(current) + 1
        if let planned = plannedRounds?.intValue { return "Round \(n) of \(planned)" }
        return "Round \(n)"
    }
}

File: apps/android/app/src/main/java/app/americanoo/android/ui/components/Components.kt
package app.americanoo.android.ui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import app.americanoo.android.ui.theme.PadelIcon
import app.americanoo.android.ui.theme.PadelTheme
import app.americanoo.android.ui.theme.Radius
import app.americanoo.android.ui.theme.Space
import app.americanoo.android.ui.theme.Tokens
import app.americanoo.data.GameStatus
import app.americanoo.data.LocalGame
import app.americanoo.data.SyncState
import app.americanoo.engine.GameState
import app.americanoo.engine.ModeId
import app.americanoo.engine.modeInfo

@Composable
fun PrimaryButton(text: String, onClick: () -> Unit, modifier: Modifier = Modifier, enabled: Boolean = true) {
    Button(
        onClick = onClick,
        enabled = enabled,
        shape = RoundedCornerShape(Radius.md),
        modifier = modifier.fillMaxWidth().heightIn(min = Tokens.touchTarget + 4.dp),
    ) { Text(text, style = MaterialTheme.typography.labelLarge) }
}

@Composable
fun SecondaryButton(text: String, onClick: () -> Unit, modifier: Modifier = Modifier, icon: PadelIcon? = null, enabled: Boolean = true) {
    OutlinedButton(
        onClick = onClick,
        enabled = enabled,
        shape = RoundedCornerShape(Radius.md),
        border = BorderStroke(1.dp, PadelTheme.colors.border),
        modifier = modifier.fillMaxWidth().heightIn(min = Tokens.touchTarget + 4.dp),
    ) {
        if (icon != null) {
            Icon(painterResource(icon.res), null, Modifier.size(20.dp), tint = PadelTheme.colors.foreground)
            Box(Modifier.size(Space.s2))
        }
        Text(text, style = MaterialTheme.typography.labelLarge, color = PadelTheme.colors.foreground)
    }
}

/** Bordered surface card, the web's `Card`. */
@Composable
fun PadelCard(modifier: Modifier = Modifier, padding: Dp = Space.s4, onClick: (() -> Unit)? = null, content: @Composable () -> Unit) {
    val shape = RoundedCornerShape(Radius.lg)
    if (onClick != null) {
        Surface(onClick = onClick, shape = shape, color = PadelTheme.colors.surface, border = BorderStroke(1.dp, PadelTheme.colors.border), modifier = modifier.fillMaxWidth()) {
            Box(Modifier.padding(padding)) { content() }
        }
    } else {
        Surface(shape = shape, color = PadelTheme.colors.surface, border = BorderStroke(1.dp, PadelTheme.colors.border), modifier = modifier.fillMaxWidth()) {
            Box(Modifier.padding(padding)) { content() }
        }
    }
}

@Composable
fun SectionTitle(text: String, trailing: String? = null) {
    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
        Text(text, style = MaterialTheme.typography.titleLarge, modifier = Modifier.weight(1f))
        if (trailing != null) Text(trailing, style = MaterialTheme.typography.bodyMedium, color = PadelTheme.colors.mutedForeground)
    }
}

/** *Live · view only*, *offline · 2 changes waiting* (FR-2.10). */
@Composable
fun StatusLine(game: LocalGame) {
    val parts = mutableListOf(if (game.game.status == GameStatus.Done) "Finished" else "Live")
    if (!game.canEdit) parts += "view only"
    when (game.sync) {
        SyncState.LocalOnly -> parts += "on this phone — shares when online"
        SyncState.Pending -> parts += "offline · ${game.pending.size} change${if (game.pending.size == 1) "" else "s"} waiting"
        SyncState.Synced -> Unit
    }
    val live = game.game.status == GameStatus.Live && game.sync == SyncState.Synced
    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(Space.s2)) {
        Box(Modifier.size(7.dp).background(if (live) PadelTheme.colors.foreground else PadelTheme.colors.mutedForeground, CircleShape))
        Text(parts.joinToString(" · "), style = MaterialTheme.typography.bodyMedium, color = PadelTheme.colors.mutedForeground)
    }
}

/** Same mapping as the web mode cards. */
fun ModeId.icon() = when (this) {
    ModeId.Americano -> PadelIcon.Rotate
    ModeId.TeamAmericano, ModeId.TeamMexicano -> PadelIcon.Pair
    ModeId.Mexicano -> PadelIcon.Ladder
    ModeId.Mixicano -> PadelIcon.Mixed
    ModeId.BeatTheBox -> PadelIcon.Box
    ModeId.UpAndDown, ModeId.TeamUpAndDown -> PadelIcon.Updown
}

@Composable
fun PadelIconView(icon: PadelIcon, size: Dp = Tokens.iconSize, modifier: Modifier = Modifier) =
    Icon(painterResource(icon.res), contentDescription = null, modifier = modifier.size(size))

fun GameState.name(id: String) = players.firstOrNull { it.id == id }?.name ?: "?"
fun GameState.team(ids: List<String>) = ids.joinToString(" & ") { name(it) }
val GameState.modeName get() = modeInfo(settings.mode).name
val GameState.roundLabel get() = "Round ${current + 1}" + (plannedRounds?.let { " of $it" } ?: "")

fun formatScore(v: Double): String = if (v % 1.0 == 0.0) v.toLong().toString() else "%.1f".format(v)

/** Chip-like selectable box used for points, rounds and formats. */
@Composable
fun RowScope.Choice(text: String, selected: Boolean, onClick: () -> Unit) {
    Surface(
        onClick = onClick,
        shape = RoundedCornerShape(Radius.md),
        color = if (selected) PadelTheme.colors.primary else PadelTheme.colors.surface,
        contentColor = if (selected) PadelTheme.colors.primaryForeground else PadelTheme.colors.foreground,
        border = BorderStroke(1.dp, if (selected) PadelTheme.colors.primary else PadelTheme.colors.border),
        modifier = Modifier.weight(1f).heightIn(min = Tokens.touchTarget),
    ) {
        Box(contentAlignment = Alignment.Center) { Text(text, style = MaterialTheme.typography.labelLarge, modifier = Modifier.padding(horizontal = Space.s2)) }
    }
}

@Composable
fun Modifier.outlined() = this.border(1.dp, PadelTheme.colors.border, RoundedCornerShape(Radius.md))
